astropy/utils/tests/test_data.py

```diff
@@ -234,6 +234,43 @@ def a_file(tmp_path):
     yield fn, contents
 
 
+def test_temp_cache(tmpdir):
+    dldir0, urlmapfn0 = _get_download_cache_locs()
+    check_download_cache(check_hashes=True)
+
+    with paths.set_temp_cache(tmpdir):
+        dldir1, urlmapfn1 = _get_download_cache_locs()
+        check_download_cache(check_hashes=True)
+        assert dldir1 != dldir0
+        assert urlmapfn1 != urlmapfn0
+
+    dldir2, urlmapfn2 = _get_download_cache_locs()
+    check_download_cache(check_hashes=True)
+    assert dldir2 != dldir1
+    assert urlmapfn2 != urlmapfn1
+    assert dldir2 == dldir0
+    assert urlmapfn2 == urlmapfn0
+
+    class Special(Exception):
```

Comment:
Maybe add a comment on why this test with `Special` is necessary?

Reply:
Can add this; I wanted to make sure both ways to exit the with statement worked. (Actually there is a third way I didn't test, through an exception raised in construction or __enter__, but this option came up when I was trying to hunt down the problem.)